src/gleam/list.gleam

```diff
@@ -2337,3 +2337,69 @@ pub fn max(
     }
   })
 }
+
+/// Take a random sample of k elements from a list using reservoir sampling.
+/// Returns an empty list if the sample size is less than or equal to 0.
+///
+/// Order is not random, only selection is.
+///
+/// ## Examples
+///
+/// ```gleam
+/// reservoir_sample([1, 2, 3, 4, 5], 3)
+/// // -> [2, 4, 5]  // A random sample of 3 items
+/// ```
+///
+pub fn sample(list: List(a), k: Int) -> List(a) {
+  case k <= 0 {
+    True -> []
+    False -> {
+      let #(reservoir, list) = split(list, k)
+
+      case length(reservoir) < k {
+        True -> reservoir
+        False -> {
+          let reservoir =
+            reservoir
+            |> map2(range(0, k - 1), _, fn(a, b) { #(a, b) })
+            |> dict.from_list
+
+          let w = float.exp(log_random() /. int.to_float(k))
+
+          do_sample(list, reservoir, k, k, w) |> dict.values
+        }
+      }
+    }
+  }
+}
+
+fn do_sample(
+  list,
+  reservoir: Dict(Int, a),
+  k,
+  index: Int,
+  w: Float,
+) -> Dict(Int, a) {
+  let skip = {
+    let assert Ok(log_result) = float.log(1.0 -. w)
+
+    log_random() /. log_result |> float.floor |> float.round
+  }
+
+  let index = index + skip + 1
+
+  case drop(list, skip) {
+    [] -> reservoir
+    [elem, ..rest] -> {
+      let reservoir = int.random(k) |> dict.insert(reservoir, _, elem)
+      let w = w *. float.exp(log_random() /. int.to_float(k))
+
+      do_sample(rest, reservoir, k, index, w)
+    }
+  }
+}
+
+fn log_random() -> Float {
```

Comment:
What does this function do? The name doesn't help me understand, so I'm not sure!

Reply:
Algo L takes the log of a uniform random value but float.random is inclusive of 0. I need to add a small value to it as log(0) is undefined (a `Result(Nil)`) to assert its not an error. I can inline it if you prefer?

Reply:
I use it in 3 places across `sample` and `sample_loop` so I thought it would be less noisy to extract it